python-package/xgboost/spark/core.py

```diff
@@ -690,8 +681,19 @@ def _fit(self, dataset):
                 num_workers,
             )
 
-        if self._repartition_needed(dataset):
-            dataset = dataset.repartition(num_workers)
+        if self._repartition_needed(dataset) or (
+            self.isDefined(self.validationIndicatorCol)
+            and self.getOrDefault(self.validationIndicatorCol)
+        ):
+            # If validationIndicatorCol defined, we always repartition dataset
+            # to balance data, because user might unionise train and validation dataset,
+            # without shuffling data then some partitions might contain only train or validation
+            # dataset.
+            # Repartition on `rand` column to avoid repartition
+            # result unbalance. Directly using `.repartition(N)` might result in some
+            # empty partitions.
+            dataset = dataset.repartition(num_workers, rand(1))
```

Comment:
I still prefer to add a parameter to control if random shuffle. Since my PR https://github.com/dmlc/xgboost/pull/8245 can detect if the training data is empty DMatrix, so we can add some log printing about data skew and how to resolve it by enabling xxx parameter. Does that make sense?